File: scripts/rust/token-contract/src/bin/balance-of.rs
use alloy::{primitives::Address, providers::ProviderBuilder, signers::local::PrivateKeySigner};
use clap::Parser;
use std::{env, str::FromStr};
use token_contract::CustomToken;
use url::Url;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Account Address
    #[arg(short, long)]
    address: Address,
}

#[tokio::main]
async fn main() -> Result<(), anyhow::Error> {
    dotenv::dotenv().ok();
    let args = Args::parse();

    let contract_str = std::env::var("TOKEN_ADDRESS")?;
    let contract = Address::from_str(&contract_str)?;
    let private_key = std::env::var("PRIVATE_KEY")?;

    // Alloy provider + signer
    let signer: PrivateKeySigner = PrivateKeySigner::from_str(&private_key)?;

    // Set up the WS transport and connect.
    let rpc_url = env::var("RPC_URL")?;
    let provider = ProviderBuilder::new()
        .wallet(signer)
        .connect_http(Url::parse(&rpc_url)?);

    let factory = CustomToken::new(contract, provider);
    let balance = factory.balanceOf(args.address).call().await?;
    println!("✅ Balance: {balance:?}");

    Ok(())
}

File: scripts/javascript/token-contract/balanceOf.js
const hre = require("hardhat");

async function main() {
  const tokenAddr = process.env.TOKEN_ADDRESS;
  if (!tokenAddr) {
    throw new Error("TOKEN_ADDRESS is not set.");
  }
  console.log("Using token:", tokenAddr);

  const token = await hre.ethers.getContractAt("CustomToken", tokenAddr);
  const decimals = await token.decimals();
  console.log("decimals:", decimals);

  let address = "0xDd3199E196BbF9A463500d5fB442FB7f78131F7a";
  // Call balanceof(address)
  let balance = await token.balanceOf(address);
  console.log("balance of", address, "is", balance);

  address = "0x69482E00b8Ab0a256E8eF99718CcD8a2C460C3f7";
  // Call balanceof(address)
  balance = await token.balanceOf(address);
  console.log("balance of", address, "is", balance);
}

main()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error(err);
    process.exit(1);
  });
